openfl-tutorials/Federated_FedProx_PyTorch_MNIST_Workflow_Tutorial.ipynb

```diff
@@ -0,0 +1,343 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Setup participants\n",
+    "aggregator_ = Aggregator()\n",
+    "aggregator_.private_attributes = {}\n",
+    "\n",
+    "# Setup collaborators with private attributes\n",
+    "collaborator_names = [f'collaborator{i}' for i in range(4)]\n",
+    "collaborators = [Collaborator(name=name) for name in collaborator_names]\n",
+    "batch_size_train = 1024\n",
+    "batch_size_test = 1024\n",
+    "log_interval = 10\n",
+    "\n",
+    "for idx, collaborator_ in enumerate(collaborators):\n",
+    "    train_images, train_labels = mnist_train.train_data, np.array(mnist_train.train_labels)\n",
+    "    train_images = torch.from_numpy(np.expand_dims(train_images, axis=1)).float()\n",
+    "    train_labels = one_hot(train_labels, 10)\n",
+    "\n",
+    "    valid_images, valid_labels = mnist_test.test_data, np.array(mnist_test.test_labels)\n",
+    "    valid_images = torch.from_numpy(np.expand_dims(valid_images, axis=1)).float()\n",
+    "\n",
+    "    collaborator_.private_attributes = {\n",
+    "            'train_loader': torch.utils.data.DataLoader(\n",
+    "                CustomDataset(train_images[idx::len(collaborators)], \n",
+    "                              train_labels[idx::len(collaborators)]), \n",
+    "                              batch_size=batch_size_train, \n",
+    "                              shuffle=True),\n",
+    "            'test_loader': torch.utils.data.DataLoader(\n",
+    "                CustomDataset(valid_images[idx::len(collaborators)], \n",
+    "                              valid_labels[idx::len(collaborators)]), \n",
+    "                              batch_size=batch_size_test, \n",
+    "                              shuffle=True)\n",
+    "    }\n",
+    "\n",
+    "local_runtime = LocalRuntime(aggregator=aggregator_, collaborators=collaborators, backend='single_process')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up work to be executed by the aggregator and the collaborators by extending `FLSpec`:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class FederatedFlow(FLSpec):\n",
+    "    def __init__(self, model=None, optimizer=None, rounds=10, **kwargs):\n",
+    "        super().__init__(**kwargs)\n",
+    "        self.model = model\n",
+    "        self.optimizer = optimizer\n",
+    "        self.rounds = rounds\n",
+    "        self.loss = 0.\n",
+    "\n",
+    "    @aggregator\n",
+    "    def start(self):\n",
+    "        print(f'Performing initialization for model')\n",
+    "        self.collaborators = self.runtime.collaborators\n",
+    "        self.private = 10\n",
```

Comment:
The `private` field isn't really needed here. It's there in the 101 tutorial to illustrate how certain attributes can be explicitly filtered out when switching to the next step of the flow.

Reply:
So if you remove the original and only keep this notebook, don't you want it here?

Reply:
It's still available in he [101_MNIST](https://github.com/securefederatedai/openfl/blob/develop/openfl-tutorials/experimental/101_MNIST.ipynb) notebook, which is not going anywhere :)

Reply:
Oh, okay, done